File: cpp/algorithms/graph/mst/prim.cpp
// Prim 最小生成树（邻接表 + 小根堆）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

int prim(int n, const vector<array<int, 3>>& edges) {
    if (n <= 1) return 0;
    vector<vector<pair<int, int>>> g(n);
    for (auto [u, v, w] : edges) {
        g[u].emplace_back(v, w);
        g[v].emplace_back(u, w);
    }
    vector<char> vis(n, 0);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    pq.emplace(0, 0);
    int tot = 0, taken = 0;
    while (!pq.empty() && taken < n) {
        auto [w, u] = pq.top();
        pq.pop();
        if (vis[u]) continue;
        vis[u] = 1;
        tot += w;
        ++taken;
        for (auto [v, wt] : g[u]) {
            if (!vis[v]) pq.emplace(wt, v);
        }
    }
    if (taken < n) throw runtime_error("graph is not connected");
    return tot;
}

int main() {
    vector<array<int, 3>> e{{0, 1, 4}, {1, 2, 3}, {0, 2, 2}};
    assert(prim(3, e) == 5);
    assert(prim(1, {}) == 0);
    try {
        prim(2, {});
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "prim OK" << endl;
    return 0;
}

File: python/algorithms/graph/mst/prim.py
"""Prim 最小生成树（邻接表 + 小根堆）。"""

from __future__ import annotations

import heapq
from typing import List, Tuple


def prim(n: int, edges: List[Tuple[int, int, int]]) -> int:
    """edges: (u, v, w) 无向；返回 MST 边权和。要求图连通。"""
    if n <= 1:
        return 0
    g: List[List[Tuple[int, int]]] = [[] for _ in range(n)]
    for u, v, w in edges:
        g[u].append((v, w))
        g[v].append((u, w))

    visited = [False] * n
    heap: List[Tuple[int, int]] = [(0, 0)]
    total = 0
    taken = 0

    while heap and taken < n:
        w, u = heapq.heappop(heap)
        if visited[u]:
            continue
        visited[u] = True
        total += w
        taken += 1
        for v, wt in g[u]:
            if not visited[v]:
                heapq.heappush(heap, (wt, v))
    if taken < n:
        raise ValueError("graph is not connected")
    return total


if __name__ == "__main__":
    e = [(0, 1, 4), (1, 2, 3), (0, 2, 2)]
    assert prim(3, e) == 5
    assert prim(1, []) == 0
    try:
        prim(2, [])
        raise AssertionError("expected ValueError")
    except ValueError:
        pass
    print("prim OK")
